lemur/certificates/service.py

```diff
@@ -1274,3 +1274,46 @@ def allowed_issuance_for_domain(common_name, extensions):
     # lemur UI copies CN as SAN (x509.DNSName). Permission check for CN might already be covered above.
     if check_permission_for_cn:
         is_authorized_for_domain(common_name)
+
+
+def send_certificate_expiration_metrics():
+    """
+    Iterate over each certificate and emit a metric for how many days until expiration.
+    """
+    success = failure = 0
+
+    certificates = database.db.session.query(Certificate).filter(
+        Certificate.expired == false(),
+        Certificate.revoked == false()
+    )
```

Comment:
Would it make sense to include a (perhaps optional) parameter limiting the query to a maximum expiry period (probably in days)? Iterating over all valid certs could potentially be a very large data set in a large Lemur deployment, and I'm guessing this metric might be most valuable for understanding certs expiring soon -- or is getting visibility into the full set of certs more the goal?

Reply:
Ability to pass in an expiry window was added as part of commit [e31ae99](https://github.com/Netflix/lemur/pull/4084/commits/e31ae99463418aab1a7ddbf0c76308e2d7129dac). The default of None will query for all active certs, while passing in a value such as 90 will only query for active certs expiring over the next 90 days. As part of this commit I moved the query logic to a separate method to make testing that logic easier.